banyand/webapp/webapp.go

```diff
@@ -0,0 +1,153 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package webapp
+
+import (
+	"fmt"
+	"io/fs"
+	"net/http"
+	"strings"
+
+	"github.com/apache/skywalking-banyandb/pkg/logger"
+	"github.com/apache/skywalking-banyandb/pkg/run"
+	"github.com/apache/skywalking-banyandb/ui"
+)
+
+type ServiceRepo interface {
+	run.Config
+	run.Service
+}
+
+var (
+	_ ServiceRepo = (*service)(nil)
+)
+
+func NewService() ServiceRepo {
+	return &service{
+		stopCh: make(chan struct{}),
+	}
+}
+
+type service struct {
+	listenAddr string
+	mux        *http.ServeMux
+	stopCh     chan struct{}
+	l          *logger.Logger
+}
+
+func (p *service) FlagSet() *run.FlagSet {
+	flagSet := run.NewFlagSet("")
+	flagSet.StringVar(&p.listenAddr, "webapp-listener-addr", ":17913", "listen addr for webapp")
+	return flagSet
+}
+
+func (p *service) Validate() error {
+	return nil
+}
+
+func (p *service) Name() string {
+	return "webapp"
+}
+
+func (p *service) PreRun() error {
+	p.l = logger.GetLogger(p.Name())
+	fSys, err := fs.Sub(ui.DistContent, "dist")
+	if err != nil {
+		return err
+	}
+	p.mux = http.NewServeMux()
+	httpFS := http.FS(fSys)
+	fileServer := http.FileServer(http.FS(fSys))
+	serveIndex := serveFileContents("index.html", httpFS)
+	p.mux.Handle("/", intercept404(fileServer, serveIndex))
```

Comment:
Then we have to mount `/api` here.

Shall we consider some web framework like `gin`?

Reply:
I didn't find any reason to import `gin` or other frameworks.

Reply:
> I didn't find any reason to import `gin` or other frameworks.

I see, currently the `http` stdlib is enough here. I'm not familiar with web-related ecosystem in Go.

Reply:
stdlib aligned our requirement which is to serve the SPA frontend and RESTFul APIs. The main repo selected a similar webapp integration path. The difference between them is that java stdlib doesn't provide a handy HTTP server and we have to adapt to the spring ecosystem.